### PR TITLE
feat: add stealth address module

### PR DESCRIPTION
**Pull request by jstinhw**
# Summary

Add stealth address module

## Change Type

- [ ] Bug Fix
- [ ] Refactor
- [x] New Feature
- [ ] Breaking Change
- [ ] Documentation Update
- [ ] Performance Improvement
- [ ] Other

# Checklist

- [ ] My code follows this project's style guidelines
- [ ] I've reviewed my own code
- [ ] I've added comments for any hard-to-understand areas
- [ ] I've updated the documentation if necessary
- [ ] My changes generate no new warnings
- [ ] I've added tests that prove my fix is effective or my feature works
- [ ] All unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

# Additional Information
Reference of stealth address module: https://hackmd.io/xgpI7DB_SPSyDOLsYGI1tA

# Branch Naming

### CODE REVIEW

**Comment by Aboudjem** [comment]
Thanks for the contribution @jstinhw ! We will review it shortly! 🚀🚀🚀

File: contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+ */
+struct StealthStorage {
+    uint256 stealthPubkey;
+    uint256 dhkey;
+    uint256 ephemeralPubkey;
+    address stealthAddress;
+    uint8 stealthPubkeyPrefix;
+    uint8 dhkeyPrefix;
+    uint8 ephemeralPrefix;
+}
```
Comment (ankurdubey521) [inline contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol:23]: Please create an interface `IStealthAddressRegistryModule` and move structs, events and errors to the interface

Comment (jstinhw) [reply contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol]: Done

File: contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol
Comment (filmakarov) [inline contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol:1]: Hey, 
First of all, thank you for contributing to Biconomy Smart Account repo!

I'm sorry I wasn't able to review the PR earlier.

File: contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        assembly {
+            let len := mload(moduleSignature)
+            mstore(add(moduleSignature, 0x01), sub(len, 1))
+            moduleSignature := add(moduleSignature, 0x01)
+        }
+
+        // 0x00: signature from spending key
+        // 0x01: aggregated signature from owner and shared secret
+        if (mode == 0x00) {
+            if (_verifySignature(dataHash, moduleSignature, msg.sender)) {
```
Comment (filmakarov) [inline contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol:121]: Can two SAs be owned by the same Stealth address?
If yes, it is required to make the resulting hash for 1271 signatures to depend on the SA's address

File: contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        address stealthAddress = _smartAccountStealth[smartAccount]
+            .stealthAddress;
+        bytes32 hash = ECDSA.toEthSignedMessageHash(dataHash);
+        if (stealthAddress == dataHash.recover(signature)) {
+            return true;
+        }
+        if (stealthAddress != hash.recover(signature)) {
+            return false;
+        }
+        return true;
```
Comment (filmakarov) [inline contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol:187]: Why not
```
if (stealthAddress != hash.recover(signature)) {
            return true;
}
return false;
```
?

Comment (filmakarov) [reply contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol:187]: Can even move
```
bytes32 hash = ECDSA.toEthSignedMessageHash(dataHash);
```
to be after 
```
if (stealthAddress == dataHash.recover(signature)) {}
```
so if `stealthAddress == dataHash.recover(signature)`, the `toEthSignedMessageHash` is not even performed

File: contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+     */
+    function _verifySignature(
+        bytes32 dataHash,
+        bytes memory signature,
+        address smartAccount
+    ) public view returns (bool) {
+        address stealthAddress = _smartAccountStealth[smartAccount]
+            .stealthAddress;
+        bytes32 hash = ECDSA.toEthSignedMessageHash(dataHash);
+        if (stealthAddress == dataHash.recover(signature)) {
```
Comment (filmakarov) [inline contracts/smart-account/modules/StealthAddressValidatoinModules/StealthAddressRegistryModule.sol:181]: Hmm. 
Maybe I'm missing something, however
if anyone can get dataHash and signature from the calldata and recover the stealth address, should it be still considered stealth then?

**Review by filmakarov — COMMENTED** [review]
Made a very high-level review, would appreciate comments.

**Pull request closed without merging**